### PR TITLE
[Proposal] WASI-crypto Proposal 

### PR DESCRIPTION
**Pull request by sonder-joker**
See #408

### CODE REVIEW

**Comment by sonder-joker** [comment]
For wasi-crypto, it can be divided into three layers.

File: include/host/wasi_crypto/symmetric/alg.h
Comment (hydai) [inline include/host/wasi_crypto/symmetric/alg.h:39]: Add newline in the end of file

File: include/host/wasi_crypto/symmetric/aeads/cipher.h
Comment (hydai) [inline include/host/wasi_crypto/symmetric/aeads/cipher.h:28]: Use `static inline constexpr` instead.

File: lib/host/wasi_crypto/symmetric/aeads/cipher.cpp
Comment (hydai) [inline lib/host/wasi_crypto/symmetric/aeads/cipher.cpp:2]: There should be a copyright line.

**Review by hydai — CHANGES_REQUESTED** [review]

File: lib/host/wasi_crypto/common/options.cpp
Comment (sonder-joker) [reply lib/host/wasi_crypto/common/options.cpp:14]: `return {std::in_place_type_t<Signatures::Options>}` seem not work, the compiler complains that`error: expected '(' for function-style cast or type construction`

File: lib/host/wasi_crypto/symmetric/eae/hkdf.cpp
Comment (hydai) [inline lib/host/wasi_crypto/symmetric/eae/hkdf.cpp:68]: Why do not create a temporary variable instead of using this dangerous function `addressOfTemporary` (This might be a undefined behaviour)?

File: lib/host/wasi_crypto/common/options.cpp
Comment (sonder-joker) [reply lib/host/wasi_crypto/common/options.cpp:14]: May `Options{std::in_place_type_t<Signatures::Options>{}}`  , a simple example here https://godbolt.org/z/ndxz9cEfj

**Review by hydai — CHANGES_REQUESTED** [review]

File: lib/host/wasi_crypto/common/options.cpp
Comment (hydai) [reply lib/host/wasi_crypto/common/options.cpp:14]: > May `return Options{std::in_place_type_t<Signatures::Options>{}}` , a simple example here https://godbolt.org/z/ndxz9cEfj

Ues your solution. `return Options{std::in_place_type_t<Signatures::Options>{}}`.

File: lib/host/wasi_crypto/symmetric/aeads/cipher.cpp
Comment (sonder-joker) [reply lib/host/wasi_crypto/symmetric/aeads/cipher.cpp:80]: Yes. Openssl have an `EVP_CIPHER_CTX_original_iv()` api to get `iv` in `EVP_CIPHER_CTX` , however, the `iv` is just zero initialization on current support algorithm(aes-gcm and chachapoly) when call init function.  I'm not sure if this is bug or by design. For future openssl may fix it, but current we can just remove it.

File: include/host/wasi_crypto/utils/evp_wrapper.h
Comment (sonder-joker) [reply include/host/wasi_crypto/utils/evp_wrapper.h:53]: Ok, I rename it to `opensslCheck`.

File: include/host/wasi_crypto/symmetric/aeads/cipher.h
Comment (sonder-joker) [reply include/host/wasi_crypto/symmetric/aeads/cipher.h:1]: I'm not sure: most helper class indicates a specific pattern(return `invalid_operation` for invalid operation) . If we need to create new algorithm(eg. blake3 for hash function) , we can simple add default behavior by inherit helper class. But if we don't need create new, merge is a good option to avoid overdesigning

File: include/host/wasi_crypto/utils/handles_manager.h
Comment (hydai) [inline include/host/wasi_crypto/utils/handles_manager.h:90]: This is confusing, please rename the `TypeID` in parameter list.

File: include/host/wasi_crypto/utils/handles_manager.h
Comment (hydai) [inline include/host/wasi_crypto/utils/handles_manager.h:116]: line 107 will find a usable handle. Do we need to check it again here?

File: include/host/wasi_crypto/utils/handles_manager.h
Comment (hydai) [inline include/host/wasi_crypto/utils/handles_manager.h:141]: `ManagerType`

File: include/host/wasi_crypto/utils/handles_manager.h
Comment (hydai) [inline include/host/wasi_crypto/utils/handles_manager.h:183]: `ManagerType`

File: include/host/wasi_crypto/utils/optional.h
Comment (hydai) [inline include/host/wasi_crypto/utils/optional.h:60]: Why not concat both conditions with `&&`?

File: include/host/wasi_crypto/utils/optional.h
Comment (hydai) [inline include/host/wasi_crypto/utils/optional.h:57]: The `enable_if_t` can be moved here to simplify the line 59-68. If you just want the `Optional` or `Function` declarations, you can use `std::declval<O>()` or `std::declval<F>()` instead. And the `forward` can also be removed in this case.

File: include/host/wasi_crypto/utils/optional.h
Comment (hydai) [inline include/host/wasi_crypto/utils/optional.h:77]: Ditto. Check the review comments in `mapAndTransposeOptional`

File: include/host/wasi_crypto/utils/optional.h
Comment (hydai) [inline include/host/wasi_crypto/utils/optional.h:94]: Ditto. Check the review comments in `mapAndTransposeOptional`

File: include/host/wasi_crypto/utils/optional.h
Comment (hydai) [inline include/host/wasi_crypto/utils/optional.h:106]: Just move the `enable_if_t` here.

File: include/host/wasi_crypto/utils/secret_key.h
Comment (hydai) [inline include/host/wasi_crypto/utils/secret_key.h:45]: `numeric_limits`.

File: include/host/wasi_crypto/utils/secret_key.h
Comment (hydai) [inline include/host/wasi_crypto/utils/secret_key.h:38]: If you want to clean all the data in vector to `0`, use `OPENSSL_cleanse` to prevent compiler optimization to remove this action.

File: include/host/wasi_crypto/utils/secret_key.h
Comment (hydai) [inline include/host/wasi_crypto/utils/secret_key.h:30]: Please add `noexcept` in this class.

**Review by hydai — CHANGES_REQUESTED** [review]

File: lib/host/wasi_crypto/symmetric/func.cpp
Comment (hydai) [inline lib/host/wasi_crypto/symmetric/func.cpp:266]: `const __wasi_opt_symmetric_key_t *`

File: lib/host/wasi_crypto/symmetric/func.cpp
Comment (hydai) [inline lib/host/wasi_crypto/symmetric/func.cpp:270]: `const __wasi_opt_options_t *`

File: lib/host/wasi_crypto/symmetric/func.cpp
Comment (hydai) [inline lib/host/wasi_crypto/symmetric/func.cpp:559]: `const uint8_t *`

File: lib/host/wasi_crypto/symmetric/func.cpp
Comment (hydai) [inline lib/host/wasi_crypto/symmetric/func.cpp:636]: `const uint8_t *`.

**Review by hydai — CHANGES_REQUESTED** [review]

File: lib/host/wasi_crypto/asymmetric_common/func.cpp
Comment (hydai) [inline lib/host/wasi_crypto/asymmetric_common/func.cpp:439]: `const uint8_t *`

**Review by hydai — CHANGES_REQUESTED** [review]

File: include/host/wasi_crypto/symmetric/eae/helper.h
Comment (sonder-joker) [reply include/host/wasi_crypto/symmetric/eae/helper.h:4]: Yes. The name is confusing but I couldn't find a better naming. Current may just use `kdf` as a better name.

File: lib/host/wasi_crypto/symmetric/eae/hkdf.cpp
Comment (sonder-joker) [reply lib/host/wasi_crypto/symmetric/eae/hkdf.cpp:68]: In previous implementation there are many place need `addressOfTemporary` (e.g `i2d` without bio), reduce temp variable that need naming. After refactoring here is the only location, I will remove it.

File: lib/host/wasi_crypto/symmetric/options.cpp
Comment (sonder-joker) [reply lib/host/wasi_crypto/symmetric/options.cpp:50]: May `insert_or_assign`?
wasi-crypto seem not have such error code

File: lib/host/wasi_crypto/symmetric/factory.cpp
Comment (sonder-joker) [reply lib/host/wasi_crypto/symmetric/factory.cpp:125]: `unlikely(!!OptKeyVariant)`?
why not  `unlikely(OptKeyVariant)`?

File: lib/host/wasi_crypto/symmetric/func.cpp
Comment (sonder-joker) [reply lib/host/wasi_crypto/symmetric/func.cpp:559]: This `RawTag` will be modify later, may use original version.

File: lib/host/wasi_crypto/symmetric/factory.cpp
Comment (hydai) [reply lib/host/wasi_crypto/symmetric/factory.cpp]: To take the boolean value.

File: lib/host/wasi_crypto/symmetric/options.cpp
Comment (hydai) [reply lib/host/wasi_crypto/symmetric/options.cpp]: I think we can use `emplace` directly? Does the spec define the behaviour about re-assign the same key? If the behaviour is overwriting the original value, then `emplace` is fine. However, if the overwriting is not allowed, then we have to use `try_emplace` and handle the existed key case.

Comment (sonder-joker) [reply lib/host/wasi_crypto/symmetric/options.cpp]: https://github.com/WebAssembly/wasi-crypto/issues/50

Comment (sonder-joker) [reply lib/host/wasi_crypto/symmetric/options.cpp]: after discussion current action should be overwrite.`insert_or_assign` is ok.

File: lib/host/wasi_crypto/symmetric/aeads/cipher.cpp
Comment (sonder-joker) [reply lib/host/wasi_crypto/symmetric/aeads/cipher.cpp]: `std::number_limit<int>::max` seem error in compare on github action. I simple cast it to size_t.

File: include/host/wasi_crypto/signatures/signatures.h
Comment (hydai) [inline include/host/wasi_crypto/signatures/signatures.h:4]: This file is `signatures.h`

File: include/host/wasi_crypto/signatures/signstate.h
Comment (hydai) [inline include/host/wasi_crypto/signatures/signstate.h:4]: This file is `signstate.h`

File: lib/host/wasi_crypto/signatures/ctx.cpp
Comment (hydai) [inline lib/host/wasi_crypto/signatures/ctx.cpp:16]: `auto &SigVariant`

File: lib/host/wasi_crypto/signatures/ctx.cpp
Comment (hydai) [inline lib/host/wasi_crypto/signatures/ctx.cpp:20]: Since you are using universal reference, you should use std::forward<decltype(Data)>(Data).

File: lib/host/wasi_crypto/signatures/ctx.cpp
Comment (hydai) [inline lib/host/wasi_crypto/signatures/ctx.cpp:34]: Since you are using universal reference, you should use std::forward<decltype(Sig)>(Sig).

File: lib/host/wasi_crypto/signatures/ctx.cpp
Comment (hydai) [inline lib/host/wasi_crypto/signatures/ctx.cpp:41]: `auto &`

File: lib/host/wasi_crypto/signatures/ctx.cpp
Comment (hydai) [inline lib/host/wasi_crypto/signatures/ctx.cpp:45]: Since you are using universal reference, you should use std::forward<decltype(SignStateVariant)>(SignStateVariant).

File: lib/host/wasi_crypto/signatures/ctx.cpp
Comment (hydai) [inline lib/host/wasi_crypto/signatures/ctx.cpp:53]: `auto &`

File: lib/host/wasi_crypto/signatures/ctx.cpp
Comment (hydai) [inline lib/host/wasi_crypto/signatures/ctx.cpp:61]: `auto &`

File: lib/host/wasi_crypto/signatures/ctx.cpp
Comment (hydai) [inline lib/host/wasi_crypto/signatures/ctx.cpp:65]: Since you are using universal reference, you should use std::forward<decltype(Signature)>(Signature).

File: lib/host/wasi_crypto/signatures/ctx.cpp
Comment (hydai) [inline lib/host/wasi_crypto/signatures/ctx.cpp:78]: `auto &`

File: lib/host/wasi_crypto/signatures/ctx.cpp
Comment (hydai) [inline lib/host/wasi_crypto/signatures/ctx.cpp:83]: Since you are using universal reference, you should use std::forward<decltype(VerificationStateVariant)>(VerificationStateVariant).

File: lib/host/wasi_crypto/signatures/ctx.cpp
Comment (hydai) [inline lib/host/wasi_crypto/signatures/ctx.cpp:91]: `auto &`

File: lib/host/wasi_crypto/signatures/ecdsa.cpp
Comment (hydai) [inline lib/host/wasi_crypto/signatures/ecdsa.cpp:104]: Why do you keep this?

**Review by hydai — CHANGES_REQUESTED** [review]

File: lib/host/wasi_crypto/signatures/ecdsa.cpp
Comment (sonder-joker) [reply lib/host/wasi_crypto/signatures/ecdsa.cpp:352]: Openssl 1.1.1 don't have such a function. I have searched but only get such a way.

**Comment by sonder-joker** [comment]
@hydai  Wait for review! If no question, I think it can merge into master.

**Comment by hydai** [comment]
> @hydai Wait for review! If no question, I think it can merge into master.

Sure. I am going to do the final review. Also, we plan to make this proposal enable in the next release. So it will get approved first, but get merged later.

**Comment by hydai** [comment]
Hi @sonder-joker 
I am going to merge this PR after the 0.10.0 release.
Could you update this branch and fix the conflicts?

**Comment by sonder-joker** [comment]
@hydai  I have solve all conflict, but it still some issues.
1. rust-sdk seem have new progress, I think it need sync while my initial implementation lack function.
2. wasi-crypto-sdk not complete enough and it need update, https://github.com/WasmEdge/WasmEdge/issues/1430 will complete it.

**Comment by hydai** [comment]
Hi @sonder-joker 
I found that your branch has conflicts after we merged the wasi-nn. Could you please rebase your branch and create a PR based on the branch `proposal/wasi-crypto`. That branch will be fixed from now to make sure your code can be merged as the first priority.

**Comment by sonder-joker** [comment]
See #1590

**Pull request closed without merging**
